### PR TITLE
Add a Jumpbox to dev stacks

### PR DESCRIPTION
**Pull request by ejrowley**
An ssh key is created based on the users public key (it assumes the
users public key is available as `~/.ssh/id_rsa.pub`), so at the moment connecting to the instance is only available for the user who is applying the jumpbox project.

In order to get the jumpbox to work you will need to ensure that you have `jq` installed and you allow your ssh key to be forwarded, details of which can be found in the README.

It's be a good idea to start with a fresh stack in order to ensure that your ssh key is being forwarded to the EC2 instance, so if you have an existing stack to destroy it and then apply the terraform to build it again.

To connect to the EC2 instance, rather than jumping onto the jumpbox and then ssh-ing on the the instance the `make jump` task will do both steps for you.

### CODE REVIEW

File: terraform/modules/common/ami/main.tf [hunk 1]
```diff
@@ -0,0 +1,25 @@
+data "aws_ami" "source" {
+  most_recent = true
+
+  filter {
+    name   = "name"
+    values = ["ubuntu/images/hvm/ubuntu-${var.ubuntu_release}-*"]
```
Comment (philandstuff) [inline terraform/modules/common/ami/main.tf]: our existing images are amazon linux - can we stick to that? (we went with it to have the baked-in optimized drivers for networking and i/o)

File: terraform/modules/common/ami/outputs.tf [hunk 2]
```diff
@@ -0,0 +1,3 @@
+output "ami_id" {
+  value = "${data.aws_ami.source.id}"
+}
```
Comment (philandstuff) [inline terraform/modules/common/ami/outputs.tf]: I don't think we need separate files for variables and outputs. @surminus explained why better than I could: https://surminus.github.io/post/terraform-structures-and-layouts/ (see section "variables and outputs")

File: terraform/modules/common/instance-profile/main.tf [hunk 3]
```diff
@@ -0,0 +1,4 @@
+resource "aws_iam_instance_profile" "profile" {
+  name = "${var.role}-${var.product}-${var.environment}-instance-profile"
+  role = "${var.role}"
+}
```
Comment (philandstuff) [inline terraform/modules/common/instance-profile/main.tf]: i don't see where this module gets used. it also looks far too small to be something worth extracting into a module

File: terraform/modules/jumpbox/main.tf [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+  type        = "ingress"
+  from_port   = 0
+  to_port     = 0
+  protocol    = "tcp"
+  cidr_blocks = ["${var.allowed_cidrs}"]
+}
+
+module "jump" {
+  source                      = "../common/instance"
```
Comment (philandstuff) [inline terraform/modules/jumpbox/main.tf]: i suspect our `terraform fmt` travis check is complaining about the alignment of all the `=` here

File: terraform/modules/jumpbox/main.tf [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  ami_id                      = "${module.ami.ami_id}"
+  ssh_key_name                = "${module.ssh_key.ssh_key_name}"
+  subnet_id                   = "${var.subnet_id}"
+  project                     = "${var.project}"
+  stack_name                  = "${var.stack_name}"
+  public_ip                   = true
```
Comment (philandstuff) [inline terraform/modules/jumpbox/main.tf]: this is probably fine for now, but i wonder if it's worth assigning an EIP?

**Comment by philandstuff** [comment]
@ejrowley can you update this PR please?

**Comment by DavidJeche** [comment]
I don't understand why we using a separate module for this when it should be something that should possibly be used in conjunction with the instances deployment. 

We could set a variable that if true the Jump box is deployed. This reduce's the amount of code & complication.

**Comment by DavidJeche** [comment]
I am just wondering about the private hosted zone that we where going to have internally. The intent was to allow us to jump to the instances more easily using DNS when inside the VPC  ?

**Comment by DavidJeche** [comment]
I think it is fine for the time being. I cannot seem to see the private DNS name attached to the instance, however this is not the most important things right. Private DNS names are luxury at the moment.

File: terraform/modules/common/ami/main.tf [hunk 6]
```diff
@@ -0,0 +1,33 @@
+## Variables
+
+variable "amazon_release" {
+  description = "The amazon release used to generate the encrypted ami"
+}
+
+variable "stack_name" {
```
Comment (idavidmcdonald) [inline terraform/modules/common/ami/main.tf]: Do we need the variable in this file given it's not used directly in it?

File: terraform/modules/common/instance/main.tf [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+  key_name = "${var.ssh_key_name}"
+
+  associate_public_ip_address = "${var.public_ip}"
+
+  security_groups             = ["${var.security_groups}"]
+  associate_public_ip_address = "${var.public_ip}"
+
+  tags {
+    Name      = "jumpbox.${var.stack_name}.dmz"
```
Comment (idavidmcdonald) [inline terraform/modules/common/instance/main.tf]: Should we have something jumpbox specific in the general instance module?

File: terraform/modules/common/instance/main.tf [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+variable "security_groups" {
+  description = "A list of security groups which will be added to the instance. If the list is empty the default securit group will be used"
+  type        = "list"
+}
+
+variable "public_ip" {
+  description = "When this is true a public ip is added to the instance"
+  default     = false
```
Comment (idavidmcdonald) [inline terraform/modules/common/instance/main.tf]: See https://www.terraform.io/docs/configuration/variables.html#booleans as I believe it is suggested to use `"false"`

File: terraform/modules/common/instance/main.tf [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+## Resources
+
+resource "aws_instance" "instance" {
+  ami           = "${var.ami_id}"
+  subnet_id     = "${var.subnet_id}"
+  instance_type = "t2.small"
```
Comment (idavidmcdonald) [inline terraform/modules/common/instance/main.tf]: Instance type could be put in a variable with t2.small as the default as it seems a bit specific to the jumpbox (I don't mind if you don't address this comment too much by the way)

File: terraform/modules/common/ssh_key/main.tf [hunk 10]
```diff
@@ -0,0 +1,18 @@
+## Variables
+
+variable "stack_name" {
+  description = "The name of the stack within which the resource resides"
+}
+
+## Resources
+
+resource "aws_key_pair" "ssh_key" {
+  key_name   = "${var.stack_name}-jumpbox-key"
```
Comment (idavidmcdonald) [inline terraform/modules/common/ssh_key/main.tf]: Do we want to have `jumpbox` in here as I believe it is supposed to be generic? Potentially instead of taking `stack_name` as a variable to the module we could take `key_name` and let the caller be responsible for naming it?

File: terraform/modules/common/ami/main.tf [hunk 11]
```diff
@@ -0,0 +1,33 @@
+## Variables
+
+variable "amazon_release" {
```
Comment (idavidmcdonald) [inline terraform/modules/common/ami/main.tf]: Is this expected to always be a version? Would appending `_version` on the end help communicate this if so? Or maybe `ami_release_version`?

File: terraform/modules/common/ami/main.tf [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+## Data sources
+
+data "aws_ami" "source" {
+  most_recent = true
+
+  filter {
+    name   = "name"
+    values = ["amzn-ami-${var.amazon_release}-amazon-ecs-optimized"]
```
Comment (idavidmcdonald) [inline terraform/modules/common/ami/main.tf]: It feels like this ami module is specific to ECS optimised instances so maybe the module should be called `ecs-ami` rather than just `ami`?

File: terraform/modules/jumpbox/main.tf [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+  security_group_id = "${aws_security_group.bastion_security_group.id}"
+
+  type        = "egress"
+  from_port   = 0
+  to_port     = 0
+  protocol    = "-1"
+  cidr_blocks = ["${var.allowed_cidrs}"]
+}
+
+module "jump" {
```
Comment (idavidmcdonald) [inline terraform/modules/jumpbox/main.tf]: I wonder if a more appropriate name for this module would be `jumpbox_instance`?

File: terraform/projects/infra-jump-instance/main.tf [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+variable "stack_name" {
+  type        = "string"
+  description = "Unique name for this collection of resources"
+  default     = "ecs-monitoring"
+}
+
+variable "enable_jumpbox" {
+  type        = "string"
+  description = "Boolean, when true a jumpbox will be deployed"
+  default     = false
```
Comment (idavidmcdonald) [inline terraform/projects/infra-jump-instance/main.tf]: https://www.terraform.io/docs/configuration/variables.html#booleans

File: terraform/projects/infra-jump-instance/README.md [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+When auto scaling groups are deployed into the private they are not accessible
+from public networks since they do not have a public IP address. In order to
+access them a jump box is in created from within this project. The instance has
+got a public IP address which can be used in order to access.
+
+There is a number of firewall rules that have been created to allow you access to all
+public + private subnets on any port. The instance is currently only allowed ssh access
+to port number 22 from any public network.
+
+In order to properly utilize this instance, you need to enable ssh-agent key forwarding.
```
Comment (idavidmcdonald) [inline terraform/projects/infra-jump-instance/README.md]: Wondering if all instructions to get this working should be in the README as it feels like someone might not know to look here?

File: terraform/modules/common/ssh_key/main.tf [hunk 16]
```diff
@@ -0,0 +1,18 @@
+## Variables
+
+variable "stack_name" {
+  description = "The name of the stack within which the resource resides"
+}
+
+## Resources
+
+resource "aws_key_pair" "ssh_key" {
```
Comment (idavidmcdonald) [inline terraform/modules/common/ssh_key/main.tf]: Is there a particular advantage to having the ssh_key as it's own file? It feels like it's closely enough tied to the jumpbox (as in we have no other need for one) that it could go into that file in the same way the security groups do

File: terraform/projects/app-ecs-instances/main.tf [hunk 17]
```diff
@@ -47,6 +41,10 @@ variable "ecs_instance_ssh_keyname" {
   default     = "ecs-monitoring-ssh-test"
 }
 
+variable "ecs_optimised_ami_version" {
```
Comment (idavidmcdonald) [inline terraform/projects/app-ecs-instances/main.tf]: This move to a version number rather than an AMI version, does it mean that the README instructions on finding the latest AMI ID is no longer so useful? (I don't know the answer)

File: setup.sh [hunk 18]
```diff
@@ -123,11 +136,19 @@ apply () {
         if [ $DEV_ENVIRONMENT = 'true' -a "$1" = 'infra-networking' ] ; then
                 import_shared_dev_route53
         fi
-        aws-vault exec ${PROFILE_NAME} -- $TERRAFORMPATH apply --var-file=$TERRAFORMTFVARS --auto-approve
+
+        # store output in a subprocess
+        exec 5>&1
+        OUTPUTS=$(aws-vault exec ${PROFILE_NAME} -- $TERRAFORMPATH apply --var-file=$TERRAFORMTFVARS --auto-approve|tee >(cat - >&5))
```
[the remark this answers is not in the document]
Comment (kentsanggds) [reply setup.sh]: not being used, could take it out, as was planning on using it to check for errors.

File: setup.sh [hunk 19]
```diff
@@ -123,11 +136,19 @@ apply () {
...
+        # TODO: capture Errors and then stop any other executions of scripts
```
[the remark this answers is not in the document]
Comment (kentsanggds) [reply setup.sh]: Not sure, I'll take out the comment as probably covers it

File: terraform/projects/infra-jump-instance/main.tf [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+  default     = "ecs-monitoring"
+}
+
+variable "stack_name" {
+  type        = "string"
+  description = "Unique name for this collection of resources"
+  default     = "ecs-monitoring"
+}
+
+variable "enable_jumpbox" {
```
[the remark this answers is not in the document]
Comment (kentsanggds) [reply terraform/projects/infra-jump-instance/main.tf]: no can take out

File: terraform/projects/infra-jump-instance/main.tf [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+  name    = "jump"
+  type    = "A"
+  ttl     = 60
+
+  records = ["${module.jumpbox.jumpbox_ip}"]
+}
+
+## Outputs
+
+output "jump_box_ip" {
```
[the remark this answers is not in the document]
Comment (kentsanggds) [reply terraform/projects/infra-jump-instance/main.tf]: ok, will make consistent

File: terraform/modules/common/ami/main.tf (hunk 6 above)
Comment (kentsanggds) [reply terraform/modules/common/ami/main.tf]: Can probably take out

File: terraform/modules/common/instance/main.tf (hunk 7 above)
Comment (kentsanggds) [reply terraform/modules/common/instance/main.tf]: I think it's only used by the jumpbox project, so maybe could move the code into jumpbox?

File: setup.sh [hunk 22]
```diff
@@ -152,6 +176,33 @@ taint() {
         aws-vault exec ${PROFILE_NAME} -- $TERRAFORMPATH taint $2
 }
 
+jumpbox() {
+        if [ $DEV_ENVIRONMENT != 'true' ] ; then
+                echo "Jumpbox is only available for dev environments"
+                return
+        fi
+
+        EC2_DATA=$(aws-vault exec $PROFILE_NAME -- aws ec2 describe-instances --filters "Name=tag:Environment,Values=$ENV" "Name=instance-state-name,Values=running")
```
[the remark this answers is not in the document]
Comment (kentsanggds) [reply setup.sh]: I'm not sure it's necessary to add additional comments, the code is fairly readable as it's split up into fairly atomic parts

File: terraform/modules/common/instance/main.tf (hunk 8 above)
Comment (kentsanggds) [reply terraform/modules/common/instance/main.tf]: can do, but there are a few places where this is not done in the codebase

File: terraform/modules/common/instance/main.tf (hunk 9 above)
Comment (kentsanggds) [reply terraform/modules/common/instance/main.tf]: Think this module should be part of jumpbox project

File: terraform/modules/common/ssh_key/main.tf (hunk 10 above)
Comment (kentsanggds) [reply terraform/modules/common/ssh_key/main.tf]: This is also only used by the jumpbox so could move to that project

File: terraform/modules/common/ami/main.tf (hunk 11 above)
Comment (kentsanggds) [reply terraform/modules/common/ami/main.tf]: Think so, though  I'm not sure if there needs to be any change as it's specific to the amazon release

File: terraform/projects/infra-jump-instance/main.tf (hunk 14 above)
Comment (kentsanggds) [reply terraform/projects/infra-jump-instance/main.tf]: Don't think this is used

File: terraform/modules/common/ami/main.tf (hunk 12 above)
Comment (kentsanggds) [reply terraform/modules/common/ami/main.tf]: I'll have a chat with Ed to see what he thinks

File: terraform/modules/jumpbox/main.tf (hunk 13 above)
Comment (kentsanggds) [reply terraform/modules/jumpbox/main.tf]: ok

File: terraform/projects/infra-jump-instance/README.md (hunk 15 above)
Comment (kentsanggds) [reply terraform/projects/infra-jump-instance/README.md]: Don't think these instructions are used, there are instructions in the main `README` that explains how to use the jumpbox

File: terraform/projects/app-ecs-instances/main.tf (hunk 17 above)
Comment (kentsanggds) [reply terraform/projects/app-ecs-instances/main.tf]: README should probably be updated? Will check with Ed

File: terraform/modules/common/ssh_key/main.tf (hunk 16 above)
Comment (kentsanggds) [reply terraform/modules/common/ssh_key/main.tf]: Could be in the jumpbox project

**Comment by idavidmcdonald** [comment]
Oh and did we ever get to the bottom of why a new jumpbox is created everytime you run make apply even if you haven't made any code changes? Would be good to avoid that (although I'd be willing to let it slide and put it as a bug ticket given how long this PR has been open)

File: setup.sh [hunk 23]
```diff
@@ -123,11 +136,17 @@ apply () {
         if [ $DEV_ENVIRONMENT = 'true' -a "$1" = 'infra-networking' ] ; then
                 import_shared_dev_route53
         fi
+
+        # store output in a subprocess
```
Comment (idavidmcdonald) [inline setup.sh]: I think this comment is no longer needed?

File: setup.sh [hunk 24]
```diff
@@ -138,6 +157,7 @@ destroy () {
                 remove_shared_dev_route53
         fi
 
+        # store output in a subprocess
```
Comment (idavidmcdonald) [inline setup.sh]: I think this one is also not needed?

File: setup.sh (hunk 22 above)
[the remark this answers is not in the document]
Comment (idavidmcdonald) [reply setup.sh]: Cool, happy to go with that then

File: terraform/projects/app-ecs-albs/main.tf [hunk 25]
```diff
@@ -128,7 +128,10 @@ resource "aws_acm_certificate" "monitoring_cert" {
 
 resource "aws_route53_record" "monitoring_cert_validation" {
   # Count matches the domain_name plus each `subject_alternative_domain`
-  count = "${1 + length(concat(aws_route53_record.prom_alias.*.fqdn, aws_route53_record.alerts_alias.*.fqdn))}"
+  # This was originally the sum of the alerts.fqdn and prom.fqdns which was nice and self documenting
```
Comment (idavidmcdonald) [inline terraform/projects/app-ecs-albs/main.tf]: Would be good to get your other PR merged and this rebased out of here as it conflicts

File: terraform/projects/infra-jump-instance/main.tf [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+## Resources
+
+module "jumpbox" {
+  source          = "../../modules/jumpbox"
+  aws_region      = "${var.aws_region}"
+  subnet_id       = "${element(data.terraform_remote_state.infra_networking.public_subnets, 0)}"
+  stack_name      = "${var.stack_name}"
+  vpc_id          = "${data.terraform_remote_state.infra_networking.vpc_id}"
+  security_groups = ["${data.terraform_remote_state.infra_security_groups.monitoring_external_sg_id}"]
+  allowed_cidrs   = "${var.jumpbox_cidrs}"
```
Comment (idavidmcdonald) [inline terraform/projects/infra-jump-instance/main.tf]: Are we allowing any IP address access to the jumpbox as opposed to just the GDS IPs here which would have been the default if we had not provided a value for `allowed_cidrs`? If so, is there a reason for that? I can't remember how cidr blocks work exactly so I might have misunderstood

File: terraform/projects/infra-jump-instance/README.md [hunk 27]
```diff
@@ -0,0 +1,56 @@
+## Project: infra-jump-instance
```
Comment (idavidmcdonald) [inline terraform/projects/infra-jump-instance/README.md]: It appears these docs have been written manually meaning running `make docs` will overwrite them. Instead they can be included at the top in a docblock of the project file and they will be produced using `make docs`.

File: Makefile [hunk 28]
```diff
@@ -49,14 +53,26 @@ apply-single: ## Apply terraform for a project, make apply-single project=<proje
 destroy-single: ## Destroy terraform for a project, make destroy-single project=<project name>
 	. ./setup.sh -d ${project}
 
-.PHONY: destroy
-destroy: ## Destroy stack
-	. ./setup.sh -d
+.PHONY: init-list
+init-list: ## Init terraform for a list of projects, make init-list list="<comma separated list of projects>"
```
Comment (idavidmcdonald) [inline Makefile]: Any chance we could put the make list stuff in as a separate PR just for ease of keeping this PR small? I haven't had a chance to look at that new code/test it

File: setup.sh (hunk 23 above)
Comment (kentsanggds) [reply setup.sh]: 👍

File: setup.sh (hunk 24 above)
Comment (kentsanggds) [reply setup.sh]: 👍

File: terraform/projects/app-ecs-albs/main.tf (hunk 25 above)
Comment (kentsanggds) [reply terraform/projects/app-ecs-albs/main.tf]: 👍

File: terraform/projects/infra-jump-instance/main.tf (hunk 26 above)
Comment (kentsanggds) [reply terraform/projects/infra-jump-instance/main.tf]: I'll check with @ejrowley on this

File: Makefile (hunk 28 above)
Comment (kentsanggds) [reply Makefile]: Sure, will create a new PR for this

File: terraform/projects/infra-jump-instance/README.md (hunk 27 above)
Comment (kentsanggds) [reply terraform/projects/infra-jump-instance/README.md]: ok

**Comment by kentsanggds** [comment]
Had a chat with @ejrowley and it appears that by adding  `security_groups` to `ignore_changes`, it should prevent the instance from being destroyed.
